keps/sig-apps/4563-evacuation-api/README.md

```diff
@@ -0,0 +1,1616 @@
+validated on [Pod Admission](#pod-admission).
+
+At this point, we know that there is a more graceful deletion of the pod available than just a PDB.
+We could ask the users making the eviction requests to create an Evacuation object instead.
+
+If the evacuation fails because the evacuator(s) do not respond, then the evacuation controller
+will proceed with the eviction. To support this, the eviction could only be allowed only at this
+point. That is, if there is a matching Evacuation object and the eviction conditions are met
+according to the [Eviction](#eviction) section.
+
+Eviction of these annotated pods would act as a PDB with a fallback to eviction if things do not go
+according to plan.
+
+This would break the eviction assumption, that eviction can proceed quickly when no PDB exists or is
+non-blocking, but it would greatly increase the safety and availability for applications.
+
+This would reduce the need to override the eviction endpoint with validating admission webhooks, as
+we can see in many projects today. They would only need a single controller responding to
+Evacuation objects.
+
+
+### Evacuation Process
+
+The following diagrams describe what the evacuation process will look like for each actor:
+
+![evacuation-process](evacuation.svg)
+
+
+### Follow-up Design Details for Kubernetes Workloads
+
+Kubernetes Workloads should be made aware of the Evacuation API to properly support the evacuation
+process.
+
+- ReplicaSet and ReplicationController partial support is considered. The full evacuation logic
+  should be implemented by a higher level workload such as Deployments. ReplicaSets are intended to
+  be a simple primitive without any sophisticated scaling logic.
+- Deployment support is considered.
+- StatefulSet support is considered.
+- DaemonSet support is considered, but will be explored as a part of the 
+  [Declarative Node Maintenance KEP](https://github.com/kubernetes/enhancements/pull/4213), as its
+  pods should not be disrupted/evacuated during normal operation.
+- Jobs/CronJobs are not yet considered, because jobs should not normally run critical workloads
+  that require evacuation and should therefore leave the node quickly and not block its drain.
+  If required, custom Evacuators can be implemented to support custom evacuation for specialized
+  types of jobs.
+
+#### ReplicaSet Controller
+
+Active response to Evacuation objects cannot be implemented without first analyzing a pod's owner
+and its capabilities. For example, not all deployment strategy configurations can simply be
+evacuated. This is also true for other owners. Therefore, it does not make sense to impose such a
+responsibility on the replication controller (e.g., by introducing new fields). Instead,
+deployments and other controllers should advertise the evacuation capabilities on ReplicaSet pods
+and implement/perform the evacuation.
+
+To facilitate the evacuation of high-level workloads and to take advantage of the Evacuation
```

Comment:
Yes, exactly. ReplicaSet changes would be minimal and we would use the existing logic unless unless Evacuation objects are present.

This would give a descheduler/custom controller the ability to  to implement any descheduling/scale down logic by creating Evacuations.